encodings/zstd/src/array.rs

```diff
@@ -0,0 +1,321 @@
+        } else {
+            Err(vortex_err!("Zstd can only encode primitive arrays"))
+        }
+    }
+
+    pub fn decompress(&self) -> VortexResult<ArrayRef> {
+        // To start, we figure out which frames we need to decompress, and with
+        // what row offset into the first such frame.
+        let slice_n_rows = self.slice_stop - self.slice_start;
+        let byte_width = self.dtype.as_ptype().byte_width();
+        let mut frame_start_row = 0;
+        let mut frame_idx_lb = 0;
+        let mut frame_idx_ub = 0;
+        let mut row_offset = 0;
+        for (i, frame_meta) in self.metadata.frames.iter().enumerate() {
+            let buf_stop = frame_start_row + usize::try_from(frame_meta.n_rows)?;
+            if frame_start_row < self.slice_start {
+                frame_idx_lb = i;
+                row_offset = self.slice_start - frame_start_row
+            }
+            if frame_start_row < self.slice_stop {
+                frame_idx_ub = i + 1
+            }
+            frame_start_row = buf_stop;
+        }
+
+        // then we actually decompress those frames
+        let frame_metas = &self.metadata.frames[frame_idx_lb..frame_idx_ub];
+        let total_uncompressed_size: usize = frame_metas
+            .iter()
+            .map(|meta| meta.uncompressed_size)
+            .sum::<u64>()
+            .try_into()?;
+
+        let mut decompressor = if let Some(dictionary) = &self.dictionary {
+            zstd::bulk::Decompressor::with_dictionary(dictionary)
+        } else {
+            zstd::bulk::Decompressor::new()
+        }?;
+
+        // we could make this empty initialized for better performance
+        let mut frames_values_bytes = vec![0; total_uncompressed_size];
+        let mut start_byte = 0;
+        for (frame, meta) in self.frames[frame_idx_lb..frame_idx_ub]
+            .iter()
+            .zip(frame_metas)
+        {
+            let stop_byte = start_byte + usize::try_from(meta.uncompressed_size)?;
+            decompressor.decompress_to_buffer(
+                frame.as_slice(),
+                &mut frames_values_bytes[start_byte..stop_byte],
+            )?;
+            start_byte = stop_byte;
+        }
+
+        // Last, we apply our offset. We need to copy since the decompressed
+        // frame start/end might not align with our slice. And we need to
+        // align the data to our (dynamic) dtype.
+        let frames_validity = self
+            .validity
+            .slice(self.slice_start - row_offset, self.slice_stop)?;
+        let frames_values_start_stop =
+            frames_validity.value_indices_for_rows(&[row_offset, row_offset + slice_n_rows])?;
+        let slice_value_bytes = frames_values_bytes
+            [frames_values_start_stop[0] * byte_width..frames_values_start_stop[1] * byte_width]
+            .to_vec();
+        let slice_values_buffer =
+            ByteBuffer::from(slice_value_bytes).aligned(Alignment::new(byte_width));
```

Comment:
We can then simplify these two into 
```suggestion
        let slice_values_buffer = frames_values_bytes.freeze().slice(
            frames_values_start_stop[0] * byte_width..frames_values_start_stop[1] * byte_width,
        );
```

Reply:
:+1: